From 383c704cd8caf09d8c9d2e8e8c9927daf818d472 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:09:06 -0400
Subject: [usb] cdc-acm: fix failed open not being detected

Message-id: <1410887463-20674-475-git-send-email-dzickus@redhat.com>
Patchwork-id: 95463
O-Subject: [RHEL7 PATCH 474/591] USB: cdc-acm: fix failed open not being detected
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 8727bf689a77a79816065e23a7a58a474ad544f9
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:43 2014 +0200

    USB: cdc-acm: fix failed open not being detected

    Fix errors during open not being returned to userspace. Specifically,
    failed control-line manipulations or control or read urb submissions
    would not be detected.

    Fixes: 7fb57a019f94 ("USB: cdc-acm: Fix potential deadlock (lockdep
    warning)")

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index 74df311..6c6928a 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -525,17 +525,17 @@ static int acm_port_activate(struct tty_port *port, struct tty_struct *tty)
  acm->control->needs_remote_wakeup = 1;
 
  acm->ctrlurb->dev = acm->dev;
- if (usb_submit_urb(acm->ctrlurb, GFP_KERNEL)) {
+ retval = usb_submit_urb(acm->ctrlurb, GFP_KERNEL);
+ if (retval) {
   dev_err(&acm->control->dev,
    "%s - usb_submit_urb(ctrl irq) failed\n", __func__);
   goto error_submit_urb;
  }
 
  acm->ctrlout = ACM_CTRL_DTR | ACM_CTRL_RTS;
- if (acm_set_control(acm, acm->ctrlout) < 0 &&
-     (acm->ctrl_caps & USB_CDC_CAP_LINE)) {
+ retval = acm_set_control(acm, acm->ctrlout);
+ if (retval < 0 && (acm->ctrl_caps & USB_CDC_CAP_LINE))
   goto error_set_control;
- }
 
  /*
   * Unthrottle device in case the TTY was closed while throttled.
@@ -545,7 +545,8 @@ static int acm_port_activate(struct tty_port *port, struct tty_struct *tty)
  acm->throttle_req = 0;
  spin_unlock_irq(&acm->read_lock);
 
- if (acm_submit_read_urbs(acm, GFP_KERNEL))
+ retval = acm_submit_read_urbs(acm, GFP_KERNEL);
+ if (retval)
   goto error_submit_read_urbs;
 
  usb_autopm_put_interface(acm->control);
@@ -564,7 +565,8 @@ error_submit_urb:
 error_get_interface:
 disconnected:
  mutex_unlock(&acm->mutex);
- return retval;
+
+ return usb_translate_errors(retval);
 }
 
 static void acm_port_destruct(struct tty_port *port)
-- 
1.7.1